### PR TITLE
Rewrite napp

### PR DESCRIPTION
**Pull request by cemsbr**
Adds refactoring to #14.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -1,30 +1,31 @@
...
 from kytos.core import KytosEvent, KytosNApp, log, rest
+from pyof.v0x01.controller2switch.flow_mod import \
+    FlowModCommand as CommonFlowModCommand
 
-from pyof.v0x01.controller2switch.flow_mod import FlowModCommand
-
-from napps.kytos.of_core.flow import Flow
-from napps.kytos.of_flow_manager import settings
+from napps.kytos.flow_manager.v0x01 import FlowSerializer as FlowSerializer10
+from napps.kytos.flow_manager.v0x04 import FlowSerializer as FlowSerializer13
```
Comment (renanrodrigo) [inline main.py]: Must be `napps.kytos.of_flow_manager`, the NApp is installed with this name.

File: v0x01.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+"""Flow serializer for OF 1.0."""
+from pyof.v0x01.common.action import ActionOutput, ActionType, ActionVlanVid
+from pyof.v0x01.controller2switch.flow_mod import FlowMod
+
+from napps.kytos.flow_manager.main import FlowSerializer as Base
```
Comment (renanrodrigo) [inline v0x01.py]: Same as above for the NApp name.
Even with that fixed, this import is not working (probably related to some relative import constraints: http://bit.ly/2g93eDE )

File: v0x04.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+from itertools import chain
+
+from pyof.foundation.network_types import HWAddress
+from pyof.v0x04.common.action import ActionOutput, ActionSetField, ActionType
+from pyof.v0x04.common.flow_instructions import InstructionType as IType
+from pyof.v0x04.common.flow_instructions import InstructionApplyAction
+from pyof.v0x04.common.flow_match import OxmOfbMatchField, OxmTLV, VlanId
+from pyof.v0x04.controller2switch.flow_mod import FlowMod
+
+from napps.kytos.flow_manager.main import FlowSerializer as Base
```
Comment (renanrodrigo) [inline v0x04.py]: Same as above for the name, also the same relative import problem.

File: openapi.yml [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+    **Warning**: *This documentation is experimental and may change soon.*
+
+
+    Manages flows for OpenFlow switches, for versions 1.0 and 1.3. Can add, list
+    and remove flows from datapaths. Currently supports a subset of available
+    matches and actions from the OpenFlow specification.
+tags:
+  - name: List
+  - name: Add
+  - name: Delete
```
Comment (macartur) [inline openapi.yml]: I guess we can remove this tags and use only one tag named Current and change the endpoints to use the methods: ['GET', 'POST', 'DELETE']. 

IMHO is wrong to use a POST method to delete information.( http://flask.pocoo.org/snippets/1/)

File: openapi.yml [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+          type: string
+          description: DPID of the target datapath.
+      produces:
+        - application/json
+      responses:
+        '200':
+          description: Operation Successful.
+          schema:
+            $ref: '#/definitions/ResponseExample'
+  /api/kytos/of_flow_manager/add-flows:
```
Comment (macartur) [inline openapi.yml]: This endpoint is not displayed.

File: openapi.yml [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+            items:
+              $ref: '#/definitions/SetFlowExample'
+      consumes:
+        - application/json
+      produces:
+        - application/json
+      responses:
+        '201':
+          description: FlowMod messages sent.
+  '/api/kytos/of_flow_manager/add-flows/{dpid}':
```
Comment (macartur) [inline openapi.yml]: This endpoint is not displayed.

File: openapi.yml [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+          type: string
+          description: DPID of the target datapath.
+      consumes:
+        - application/json
+      produces:
+        - application/json
+      responses:
+        '201':
+          description: FlowMod messages sent.
+  /api/kytos/of_flow_manager/del-flows:
```
Comment (macartur) [inline openapi.yml]: This endpoint is not displayed.

File: openapi.yml [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            items:
+              $ref: '#/definitions/DeleteFlowExample'
+      consumes:
+        - application/json
+      produces:
+        - application/json
+      responses:
+        '202':
+          description: FlowMod messages sent.
+  '/api/kytos/of_flow_manager/del-flows/{dpid}':
```
Comment (macartur) [inline openapi.yml]: This endpoint is not displayed.

File: openapi.yml (hunk 4 above)
Comment (renanrodrigo) [reply openapi.yml]: @diraol and I do think this should be a DELETE endpoint. However, we cannot send a JSON body with DELETE (https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE), and the flows are being removed according to their attributes, sent via JSON.

File: openapi.yml [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+      tags:
+        - List
+      summary: Retrieve a list of all flows from all known datapaths.
+      produces:
+        - application/json
+      responses:
+        '200':
+          description: Operation Successful.
+          schema:
+            $ref: '#/definitions/ResponseExample'
```
[the remark this answers is not in the document]
Comment (renanrodrigo) [reply openapi.yml]: The OpenAPI file was created using v2.0. I changed it to 3.0 to check incompatibility. If they are not displaying, then they are not compatible...
1. Do you think keeping it as 2.0 would solve this? Can the NApps Server handle that? or
2. Do you think it should be rewritten to match the 3.0 syntax?

File: main.py [hunk 10]
```diff
@@ -37,122 +36,67 @@ def shutdown(self):
...
-            switch = self.controller.get_switch_by_dpid(switch_dpid)
-            flows = {}
-            for flow in switch.flows:
-                flow = (flow.as_dict()['flow'])
-                flow_id = flow.pop('self.id', 0)
-                flows[flow_id] = flow
-            switch_flows[switch_dpid] = flows
+        for switch in switches:
+            serializer = self._get_serializer(switch)
+            flows_dict = [serializer.as_dict(flow) for flow in switch.flows]
```
Comment (renanrodrigo) [inline main.py]: `as_dict` -> `to_dict`

File: serializers/v0x04.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        tlv = OxmTLV()
+        tlv.oxm_field = OxmOfbMatchField.OFPXMT_OFB_VLAN_VID
+        oxm_value = vlan_id | VlanId.OFPVID_PRESENT
+        tlv.oxm_value = oxm_value.to_bytes(2, 'big')
+        return tlv
+
+    def to_dict(self, flow_stats):
+        """Return a dictionary created from input 1.3 switch's flows."""
+        flow_dict = {field: data.value
+                     for field, data in vars(flow_stats).items()}
```
Comment (renanrodrigo) [inline serializers/v0x04.py]: Here we need to specify which items we want: `field` must be in `flow_attributes`.

I'm getting: 
`Match object has no attribute 'value'`

Comment (renanrodrigo) [reply serializers/v0x04.py]: (same for v0x01, in line 58)

**Pull request merged**